### PR TITLE
ENH: Update formulaic to 0.6

### PR DESCRIPTION
**Pull request by bashtage**
Update to latest release
Add __future__ module

closes #521

### CODE REVIEW

File: linearmodels/iv/_utility.py [hunk 1]
```diff
@@ -78,12 +83,27 @@ class IVFormulaParser:
...
         self._eval_env = eval_env
+        if not context:
+            self._context = capture_context(context=self._eval_env)
+        else:
+            self._context = context
         self._components: dict[str, str] = {}
         self._parse()
+        if FORMULAIC_GTE_0_6 and not future_ordering():
+            self._formulaic_kwargs = dict(_ordering="sort")
+            self._model_matrix_kwargs = dict(cluster_by="numerical_factors")
```
Comment (matthewwardrop) [inline linearmodels/iv/_utility.py:104]: Note: This was *not* the default behaviour from before. This was added in 0.6 to allow matching `patsy`'s behaviour. To match formulaic < 0.6, omit the `cluster_by` option here.

You can choose whether you want the future behaviour to cluster like patsy or not. I can see good arguments for both, but I lean to accepting the default of `cluster_by="none"`.

File: linearmodels/iv/_utility.py [hunk 2]
```diff
@@ -78,12 +83,27 @@ class IVFormulaParser:
     The general structure of a formula is `dep ~ exog + [endog ~ instr]`
     """
 
-    def __init__(self, formula: str, data: DataFrame, eval_env: int = 2):
+    def __init__(
+        self,
+        formula: str,
+        data: DataFrame,
+        eval_env: int = 2,
+        context: LayeredMapping | None = None,
```
Comment (matthewwardrop) [inline linearmodels/iv/_utility.py:91]: nit: Context can just be `Mapping[str, Any]`. Internally we map this to a `LayeredMapping` for better isolation and attribution.

File: linearmodels/panel/model.py [hunk 3]
```diff
@@ -141,12 +145,28 @@ class PanelFormulaParser:
     The general structure of a formula is `dep ~ exog`
     """
 
-    def __init__(self, formula: str, data: PanelDataLike, context: int = 2) -> None:
+    def __init__(
+        self,
+        formula: str,
+        data: PanelDataLike,
+        eval_env: int = 2,
+        context: LayeredMapping | None = None,
```
Comment (matthewwardrop) [inline linearmodels/panel/model.py:153]: Same here; you can just use `Mapping[str, Any]`.

File: linearmodels/panel/model.py [hunk 4]
```diff
@@ -141,12 +145,28 @@ class PanelFormulaParser:
...
-        self._context = context
+        self._eval_env = eval_env
+        if not context:
+            self._context = capture_context(eval_env)
+        else:
+            self._context = context
         self._dependent = self._exog = None
+        if FORMULAIC_GTE_0_6 and not future_ordering():
+            self._formulaic_kwargs = dict(_ordering="sort")
+            self._model_matrix_kwargs = dict(cluster_by="numerical_factors")
```
Comment (matthewwardrop) [inline linearmodels/panel/model.py:166]: Same comment as above re: default behaviour.

File: linearmodels/panel/results.py [hunk 5]
```diff
@@ -472,15 +474,19 @@ def resids(self) -> Series:
         return Series(self._resids.squeeze(), index=self._index, name="residual")
 
     def _out_of_sample(
-        self, exog: ArrayLike | None, data: DataFrame | None, missing: bool
+        self,
+        exog: ArrayLike | None,
+        data: DataFrame | None,
+        missing: bool,
+        context: LayeredMapping | None = None,
```
Comment (matthewwardrop) [inline linearmodels/panel/results.py:481]: Ditto about `LayeredMapping` -> `Mapping[str, Any]`.

File: linearmodels/system/model.py [hunk 6]
```diff
@@ -148,7 +150,8 @@ def __init__(
         formula: Mapping[str, str] | str,
         data: DataFrame,
         weights: Mapping[str, ArrayLike] | None = None,
-        eval_env: int = 6,
+        eval_env: int = 2,
+        context: LayeredMapping | None = None,
```
Comment (matthewwardrop) [inline linearmodels/system/model.py:154]: Ditto, just to prevent  this being missed.

**Review by matthewwardrop — APPROVED** [review]
Left a few small comments, but otherwise looks good to me. Feel free to disregard if you feel them unhelpful :).

File: linearmodels/iv/_utility.py (hunk 2 above)
Comment (bashtage) [reply linearmodels/iv/_utility.py]: Thanks. Changed.

File: linearmodels/iv/_utility.py (hunk 1 above)
Comment (bashtage) [reply linearmodels/iv/_utility.py]: Thanks. Changed.

**Pull request merged**